doc/src/manual/methods.md

```diff
@@ -515,6 +515,226 @@ julia> wait(schedule(t, 1))
 "definition for Int"
 ```
 
+## Design Patterns with Parametric Methods
+
+
+While complex dispatch logic is not required for performance or usability,
+sometimes it can be the best way to express some algorithm.
+Here are a few common design patterns that come up sometimes when using dispatch in this way.
+
+### Extracting the type parameter from a super-type
+
+
+Here is the correct code template for returning the element-type `T`
+of any arbitrary subtype of `AbstractArray`:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{<:AbstractArray{T}}) where {T} = T
+```
+using so-called triangular dispatch.
+
+Another way, which used to be the only correct way before the advent of
+triangular dispatch, is:
+
+```julia
+abstract AbstractArray{T, N}
+eltype(::Type{AbstractArray}) = Any
+eltype{T}(::Type{AbstractArray{T}}) = T
+eltype{T, N}(::Type{AbstractArray{T, N}}) = T
+eltype{A<:AbstractArray}(::Type{A}) = eltype(supertype(A))
+```
```

Comment:
The proper way to write this in the new type (v0.6) system (with examples of which types would trigger each) is:
```julia
eltype(::Type{AbstractArray{T, N} where {T<:S, N<:M}}) where {M, S} = Any
  # eltype(Array) = Any
  # eltype(Array{T} where T <: Integer) = Any
eltype(::Type{AbstractArray{T, N} where {T<:S}}) where {N, S} = Any
  # eltype(Array{T, 1} where T <: Integer) = Any
eltype(::Type{AbstractArray{T, N} where {N<:M}}) where {M, T} = T
  # eltype(Array{Integer}) = Integer
  # eltype(Array{Integer, N} where N<:Integer) = Integer
eltype(::Type{AbstractArray{T, N}}) where {T, N} = T
  # eltype(Array{Integer, 3}) = Integer
eltype(::Type{A}) where {A <: AbstractArray} = eltype(supertype(A))
```